sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/RebaseDateTimeSuite.scala

```diff
@@ -0,0 +1,352 @@
+        .setTimeZone(TimeZoneUTC)
+        .setInstant(Math.multiplyExact(days, MILLIS_PER_DAY))
+        .build()
+      val localDate = LocalDate.of(
+        utcCal.get(Calendar.YEAR),
+        utcCal.get(Calendar.MONTH) + 1,
+        // The number of days will be added later to handle non-existing
+        // Julian dates in Proleptic Gregorian calendar.
+        // For example, 1000-02-29 exists in Julian calendar because 1000
+        // is a leap year but it is not a leap year in Gregorian calendar.
+        1)
+        .plusDays(utcCal.get(Calendar.DAY_OF_MONTH) - 1)
+      Math.toIntExact(localDate.toEpochDay)
+    }
+
+    val start = rebaseGregorianToJulianDays(
+      localDateToDays(LocalDate.of(1, 1, 1)))
+    val end = rebaseGregorianToJulianDays(
+      localDateToDays(LocalDate.of(2030, 1, 1)))
+
+    var days = start
+    while (days < end) {
+      assert(rebaseJulianToGregorianDays(days) === refRebaseJulianToGregorianDays(days))
+      days += 1
+    }
+  }
+
+  test("SPARK-31328: rebasing overlapped timestamps during daylight saving time") {
+    Seq(
+      LA.getId -> Seq("2019-11-03T08:00:00Z", "2019-11-03T08:30:00Z", "2019-11-03T09:00:00Z"),
+      "Europe/Amsterdam" ->
+        Seq("2019-10-27T00:00:00Z", "2019-10-27T00:30:00Z", "2019-10-27T01:00:00Z")
+    ).foreach { case (tz, ts) =>
+      withDefaultTimeZone(getZoneId(tz)) {
+        ts.foreach { str =>
+          val micros = instantToMicros(Instant.parse(str))
+          assert(rebaseGregorianToJulianMicros(micros) === micros)
+          assert(rebaseJulianToGregorianMicros(micros) === micros)
+        }
+      }
+    }
+  }
+
+  test("validate rebase records in JSON files") {
+    import collection.JavaConverters._
+    Seq(
+      "gregorian-julian-rebase-micros.json",
+      "julian-gregorian-rebase-micros.json").foreach { json =>
+      withClue(s"JSON file = $json") {
+        val rebaseRecords = loadRebaseRecords(json)
+        rebaseRecords.asScala.foreach { case (_, rebaseRecord) =>
+          assert(rebaseRecord.switches.size === rebaseRecord.diffs.size)
+          // Check ascending order of switches values
+          assert(rebaseRecord.switches.toSeq === rebaseRecord.switches.sorted.toSeq)
+        }
+      }
+    }
+  }
+
+  test("optimization of micros rebasing - Gregorian to Julian") {
+    outstandingZoneIds.foreach { zid =>
+      withClue(s"zone id = $zid") {
+        withDefaultTimeZone(zid) {
+          val start = instantToMicros(LocalDateTime.of(1, 1, 1, 0, 0, 0)
+            .atZone(zid)
+            .toInstant)
+          val end = instantToMicros(LocalDateTime.of(2100, 1, 1, 0, 0, 0)
+            .atZone(zid)
+            .toInstant)
+          var micros = start
+          do {
+            val rebased = rebaseGregorianToJulianMicros(zid, micros)
+            val rebasedAndOptimized = rebaseGregorianToJulianMicros(micros)
+            assert(rebasedAndOptimized === rebased)
+            micros += MICROS_PER_MONTH
```

Comment:
can we use a random step (20 days to 40 days) to increase test coverage?

Reply:
done